Improve companion doc (#4227)
transloadit/uppy: website/src/docs/companion.md

We have [a detailed guide on running Companion in Kubernetes](https://github.com/transloadit/uppy/blob/main/packages/%40uppy/companion/KUBERNETES.md) for you, that’s how we run our example server at <https://companion.uppy.io>.

> ### Running many instances

Reviewer: Since we've been running Companion for many customers for a while, it would be great if we could add some data on when another instance is needed? Do we have any insight and how much a single instance can handle?

Reviewer: Can you copy and edit this message: https://community.transloadit.com/t/performance-testing-of-uppy-companion/16157/3?u=merlijn

Author: I think in general for Node.js server apps it's always recommended to run more than one instance on each server, because if some code in one request blocks the event loop, all other requests get blocked too if only one instance. When uploading a lot of tiny files in parallel, the event loop could get saturated fast. I think the number of instances to run would usually be limited the amount of system memory as well as number of CPUs.

Two ways of running many concurrent Companion instances.

#### Separate domain names

> One option is to run many instances with each instance having its own (sub)domain name. With this setup, the Companion client in Uppy will direct requests to companion in a way so that all requests of a particular upload will always be sent to the same instance. You would then also typically configure a single instance (one domain name) to handle all OAuth authentication requests, so that you only need to specify a single OAuth callback URL. See also `oauthDomain` and `validHosts`

Reviewer 2: ```suggestion
One option is to run many instances with each instance having its own unique endpoint. This could be on separate ports, (sub)domain names, or IPs. With this setup, the Companion client in Uppy will direct requests to companion in a way so that all requests of a particular upload will always be sent to the same endpoint, and hence process. You would then also typically configure a single instance (one endpoint) to handle all OAuth authentication requests, so that you only need to specify a single OAuth callback URL. See also `oauthDomain` and `validHosts`
```

> One option is to run many instances with each instance having its own unique endpoint. This could be on separate ports, (sub)domain names, or IPs. With this setup, the Companion client in Uppy will direct requests to companion in a way so that all requests of a particular upload will always be sent to the same endpoint, and hence process. You would then also typically configure a single instance (one endpoint) to handle all OAuth authentication requests, so that you only need to specify a single OAuth callback URL. See also `oauthDomain` and `validHosts`

Reviewer: How does this work in Uppy where you can only set a single `companionUrl`?

Author: I think this code https://github.com/transloadit/uppy/blob/bffd5dc9ca1d945ac2c566ca0328d5a174cc4cce/packages/%40uppy/companion-client/src/RequestClient.js#L79

 and this https://github.com/transloadit/uppy/blob/bffd5dc9ca1d945ac2c566ca0328d5a174cc4cce/packages/%40uppy/companion-client/src/RequestClient.js#L57

I'll try to explain it

Author: side note: sending i-am from companion is enabled by setting COMPANION_SELF_ENDPOINT. However api2 does not set this env variable, so it does not use this functionality (any more).

However the Transloadit Uppy plugin has its own logic where it overrides companionUrl as soon as it receives a response from the created assembly:

https://github.com/transloadit/uppy/blob/5c0967fcdea7036517d0c480a6ff7b005e67d5b6/packages/%40uppy/transloadit/src/index.js#L163

So this is how Uppy/Transloadit pins each upload to a single companion instance.

#### Behind a load balancer

The other option is to set up a load balancer in front of many Companion instances. Then Uppy’s companion client will only see a single endpoint and send all requests to the associated load balancer, which will then distribute them between Companion instances. The companion instances will then coordinate their messages and events over Redis so that any instance can serve the client’s requests. Note that sticky sessions are **not** needed with this setup. Here are some requirements for this setup:

* The instances need to be connected to the same Redis server.
* You need to set `COMPANION_SECRET` to the same value on both servers.
* if you use the `companionKeysParams` feature (Transloadit), you also need `COMPANION_PREAUTH_SECRET` to be the same on each instance.
* All other configuration needs to be the same, except if you’re running many instances on the same machine, then `COMPANION_PORT` should be different for each instance.

> For more information about this setup [see this issue](https://github.com/transloadit/uppy/issues/3538).

Author: no, I think i've pretty much summed up what we learned in that issue

> For more information about this setup [see this issue](https://github.com/transloadit/uppy/issues/3538).
> 

Author: ```suggestion
```

We have [a detailed guide on running Companion in Kubernetes](https://github.com/transloadit/uppy/blob/main/packages/%40uppy/companion/KUBERNETES.md) for you, that’s how we run our example server at <https://companion.uppy.io>.

### Running many instances

Two ways of running many concurrent Companion instances.

In our experience Companion will saturate network interface cards before other resources on commodity virtual servers (`c5d.2xlarge` for instance). We recommend running at least two instances in production, so that if the Node.js event loop gets blocked by one or more requests (due to a bug or spike in traffic), it doesn’t also block or slow down all other requests as well (due to Node.js single threaded nature.) As an example for scale, one OSS enterprise customer of Transloadit that self-hosts Companion to power an education service that is deployed by virtually all universities globally, deploys 7 Companion instances, their earlier solution ran on 35 instances I believe.

But as always it depends and your mileage may vary, so we recommend to add observability. You can let Prometheus crawl the /metrics endpoint and graph that with Grafana for instance.

> #### Separate endpoints

Author: I think the term "sticky sessions" is normally used with load balancers. so using this term here implies that this option also involves using  a load balancer and sticking http cookies to upstream servers in the lb

> #### Separate endpoints

Reviewer: Good point, I got confused about that for a moment. New suggestion:

```suggestion
#### Using unique endpoints
```

One option is to run many instances with each instance having its own unique endpoint. This could be on separate ports, (sub)domain names, or IPs. With this setup, you can either
1. Implement your own logic that will direct each upload to a specific Companion endpoint by setting the `companionUrl` option

> 2. Setting the Companion option `COMPANION_SELF_ENDPOINT`. This option will cause Companion to respond with a special `i-am` HTTP header containing the value from `COMPANION_SELF_ENDPOINT`. When Uppy’s Companion client sees this header, it will pin all requests for the upload to this endpoint.

Reviewer: Companion client is a internal tool, and might be confusing lingo for people. Better to refer to Uppy. 

```suggestion
2. Setting the Companion option `COMPANION_SELF_ENDPOINT`. This option will cause Companion to respond with a `i-am` HTTP header containing the value from `COMPANION_SELF_ENDPOINT`. When Uppy’s sees this header, it will pin all requests for the upload to this endpoint.
```

We have [a detailed guide on running Companion in Kubernetes](https://github.com/transloadit/uppy/blob/main/packages/%40uppy/companion/KUBERNETES.md) for you, that’s how we run our example server at <https://companion.uppy.io>.

### Running many instances

> In our experience Companion will saturate network interface cards before other resources on commodity virtual servers (`c5d.2xlarge` for instance). We recommend running at least two instances in production, so that if the Node.js event loop gets blocked by one or more requests (due to a bug or spike in traffic), it doesn’t also block or slow down all other requests as well (as Node.js is single threaded).
> 
> As an example for scale, one enterprise customer of Transloadit, who self-hosts Companion to power an education service that is used by many universities globally, deploys 7 Companion instances. Their earlier solution ran on 35 instances I believe.
> 
> Your mileage may vary, so we recommend to add observability. You can let Prometheus crawl the `/metrics` endpoint and graph that with Grafana for instance.

Reviewer: ```suggestion
We recommend running at least two instances in production, so that if the Node.js event loop gets blocked by one or more requests (due to a bug or spike in traffic), it doesn’t also block or slow down all other requests as well (as Node.js is single threaded).

As an example for scale, one enterprise customer of Transloadit, who self-hosts Companion to power an education service that is used by many universities globally, deploys 7 Companion instances. Their earlier solution ran on 35 instances. In our general experience Companion will saturate network interface cards before other resources on commodity virtual servers (`c5d.2xlarge` for instance).

Your mileage may vary, so we recommend to add observability. You can let Prometheus crawl the `/metrics` endpoint and graph that with Grafana for instance.
```
